{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5809b8f8",
   "metadata": {},
   "source": [
    "T1D vs T2D 간 beta-cell 유전자 발현 차이"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "4f5387ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.sparse import issparse"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a0205d",
   "metadata": {},
   "source": [
    "Data loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d32f40df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import gzip\n",
    "# with gzip.open(\"C:/Users/user/Desktop/GSE211799_adata_atlas.h5ad.gz\", 'rb') as f:\n",
    "#     adata = ad.read_h5ad(f)\n",
    "\n",
    "# Raw data 구조 및 구성 파악\n",
    "adata = sc.read_h5ad('C:/Users/user/Desktop/T1D/GSE211799_adata_atlas.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "60708c14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(301796, 31706)\n",
      "AnnData object with n_obs × n_vars = 301796 × 31706\n",
      "    obs: 'study_sample', 'study', 'file', 'reference', 'size_factors_sample', 'phase_cyclone', 's_cyclone', 'g2m_cyclone', 'g1_cyclone', 'sex', 'ins_score', 'ins_high', 'gcg_score', 'gcg_high', 'sst_score', 'sst_high', 'ppy_score', 'ppy_high', 'cell_filtering', 'age', 'strain', 'tissue', 'technique', 'study_sample_design', 'cell_type', 'cell_type_multiplet', 'cell_subtype', 'cell_subtype_multiplet', 'design', 'size_factors_integrated', 'pre_cell_type_unified', 'pre_cell_type_original', 'study_parsed', 'cell_type_parsed', 'low_q', 'BETA-DATA_leiden_r1.5', 'BETA-DATA_leiden_r20', 'BETA-DATA_hc_gene_programs', 'BETA-DATA_hc_gene_programs_parsed', 'BETA-DATA_leiden_r1.5_parsed', 'BETA-DATA_leiden_r1.5_parsed_const', 'CXG-DATA_n_genes', 'CXG-DATA_mt_frac', 'CXG-DATA_doublet_score', 'CXG-DATA_log10_n_counts', 'CXG-DATA_age_approxDays', 'CXG-DATA_cell_subtype_immune_reannotatedIntegrated', 'CXG-DATA_cell_subtype_endothelial_reannotatedIntegrated', 'CXG-DATA_emptyDrops_LogProb_scaled', 'CXG-DATA_diabetes_model', 'CXG-DATA_chemical_stress', 'CXG-DATA_GEO_accession', 'CXG-DATA_sex_annotation', 'cell_type_integrated_v2', 'cell_type_integrated_v2_parsed'\n",
      "    var: 'gene_symbol', 'used_integration', 'gene_symbol_original_matched', 'CXG-DATA_feature_is_filtered', 'CXG-DATA_present_Fltp_2y', 'CXG-DATA_present_Fltp_adult', 'CXG-DATA_present_Fltp_P16', 'CXG-DATA_present_NOD', 'CXG-DATA_present_NOD_elimination', 'CXG-DATA_present_spikein_drug', 'CXG-DATA_present_embryo', 'CXG-DATA_present_VSG', 'CXG-DATA_present_STZ', 'gene_symbol_FINAL'\n",
      "    uns: 'BETA-DATA_hc_gene_programs_parsed_colors', 'BETA-DATA_hc_gene_programs_parsed_order', 'BETA-DATA_leiden_r1.5_parsed_colors', 'BETA-DATA_leiden_r1.5_parsed_order', 'cell_type_integrated_v2_colors', 'cell_type_integrated_v2_parsed_order', 'field_descriptions', 'study_colors', 'study_order', 'study_parsed_colors', 'study_parsed_order'\n",
      "    obsm: 'BETA-DATA_X_umap', 'BETA-DATA_X_umap_opt', 'X_integrated', 'X_umap'\n",
      "                                                             study_sample  \\\n",
      "index                                                                       \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination  NOD_elimination_SRR7610301   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination  NOD_elimination_SRR7610298   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination  NOD_elimination_SRR7610301   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                              VSG_MUC13640   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination  NOD_elimination_SRR7610296   \n",
      "\n",
      "                                                         study        file  \\\n",
      "index                                                                        \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination  NOD_elimination  SRR7610301   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination  NOD_elimination  SRR7610298   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination  NOD_elimination  SRR7610301   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                            VSG    MUC13640   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination  NOD_elimination  SRR7610296   \n",
      "\n",
      "                                               reference  size_factors_sample  \\\n",
      "index                                                                           \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination      False             1.815367   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination      False             1.857232   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination      False             1.109190   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                    False             0.159150   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination      False             0.344788   \n",
      "\n",
      "                                              phase_cyclone  s_cyclone  \\\n",
      "index                                                                    \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination            G1      0.006   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination            G1      0.000   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination            G1      0.009   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                          G1      0.000   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination            G1      0.035   \n",
      "\n",
      "                                               g2m_cyclone  g1_cyclone  \\\n",
      "index                                                                    \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination        0.003       1.000   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination        0.027       1.000   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination        0.086       0.999   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                      0.491       0.995   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination        0.002       0.998   \n",
      "\n",
      "                                                  sex  ...  \\\n",
      "index                                                  ...   \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination  female  ...   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination  female  ...   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination  female  ...   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                  male  ...   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination  female  ...   \n",
      "\n",
      "                                               CXG-DATA_age_approxDays  \\\n",
      "index                                                                    \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                    112.0   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                     98.0   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                    112.0   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                                  119.0   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                     56.0   \n",
      "\n",
      "                                               CXG-DATA_cell_subtype_immune_reannotatedIntegrated  \\\n",
      "index                                                                                               \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                                                NaN    \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                                                NaN    \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                                                NaN    \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                                                               DC    \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                                                NaN    \n",
      "\n",
      "                                               CXG-DATA_cell_subtype_endothelial_reannotatedIntegrated  \\\n",
      "index                                                                                                    \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                                                NaN         \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                                                NaN         \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                                           capilary         \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                                                              NaN         \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                                                NaN         \n",
      "\n",
      "                                               CXG-DATA_emptyDrops_LogProb_scaled  \\\n",
      "index                                                                               \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                            0.639223   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                            0.682840   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                            0.768452   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                                          0.918929   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                            0.977167   \n",
      "\n",
      "                                               CXG-DATA_diabetes_model  \\\n",
      "index                                                                    \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                  T1D_NOD   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                  T1D_NOD   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                  T1D_NOD   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                  T2D_db/db-treated_VSG   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination      T1D_NOD_prediabetic   \n",
      "\n",
      "                                               CXG-DATA_chemical_stress  \\\n",
      "index                                                                     \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                       NaN   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                       NaN   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                       NaN   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                                     NaN   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                       NaN   \n",
      "\n",
      "                                               CXG-DATA_GEO_accession  \\\n",
      "index                                                                   \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination               GSE117770   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination               GSE117770   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination               GSE117770   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                             GSE174194   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination               GSE117770   \n",
      "\n",
      "                                               CXG-DATA_sex_annotation  \\\n",
      "index                                                                    \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination             ground-truth   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination             ground-truth   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination             ground-truth   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                           ground-truth   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination             ground-truth   \n",
      "\n",
      "                                              cell_type_integrated_v2  \\\n",
      "index                                                                   \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                  ductal   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                   gamma   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination             endothelial   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                                immune   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination      stellate_quiescent   \n",
      "\n",
      "                                              cell_type_integrated_v2_parsed  \n",
      "index                                                                         \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                         ductal  \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                          gamma  \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                    endothelial  \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                                       immune  \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                    stellate q.  \n",
      "\n",
      "[5 rows x 55 columns]\n",
      "                   gene_symbol  used_integration gene_symbol_original_matched  \\\n",
      "EID                                                                             \n",
      "ENSMUSG00000000001       Gnai3             False                        Gnai3   \n",
      "ENSMUSG00000000003        Pbsn             False                         Pbsn   \n",
      "ENSMUSG00000000028       Cdc45             False                        Cdc45   \n",
      "ENSMUSG00000000031         H19             False                          H19   \n",
      "ENSMUSG00000000037       Scml2             False                        Scml2   \n",
      "\n",
      "                    CXG-DATA_feature_is_filtered  CXG-DATA_present_Fltp_2y  \\\n",
      "EID                                                                          \n",
      "ENSMUSG00000000001                         False                      True   \n",
      "ENSMUSG00000000003                         False                      True   \n",
      "ENSMUSG00000000028                         False                      True   \n",
      "ENSMUSG00000000031                         False                      True   \n",
      "ENSMUSG00000000037                         False                      True   \n",
      "\n",
      "                    CXG-DATA_present_Fltp_adult  CXG-DATA_present_Fltp_P16  \\\n",
      "EID                                                                          \n",
      "ENSMUSG00000000001                         True                       True   \n",
      "ENSMUSG00000000003                         True                       True   \n",
      "ENSMUSG00000000028                         True                       True   \n",
      "ENSMUSG00000000031                         True                       True   \n",
      "ENSMUSG00000000037                         True                       True   \n",
      "\n",
      "                    CXG-DATA_present_NOD  CXG-DATA_present_NOD_elimination  \\\n",
      "EID                                                                          \n",
      "ENSMUSG00000000001                  True                              True   \n",
      "ENSMUSG00000000003                  True                              True   \n",
      "ENSMUSG00000000028                  True                              True   \n",
      "ENSMUSG00000000031                  True                              True   \n",
      "ENSMUSG00000000037                  True                              True   \n",
      "\n",
      "                    CXG-DATA_present_spikein_drug  CXG-DATA_present_embryo  \\\n",
      "EID                                                                          \n",
      "ENSMUSG00000000001                           True                     True   \n",
      "ENSMUSG00000000003                           True                     True   \n",
      "ENSMUSG00000000028                           True                     True   \n",
      "ENSMUSG00000000031                           True                     True   \n",
      "ENSMUSG00000000037                           True                     True   \n",
      "\n",
      "                    CXG-DATA_present_VSG  CXG-DATA_present_STZ  \\\n",
      "EID                                                              \n",
      "ENSMUSG00000000001                  True                  True   \n",
      "ENSMUSG00000000003                  True                  True   \n",
      "ENSMUSG00000000028                  True                  True   \n",
      "ENSMUSG00000000031                  True                  True   \n",
      "ENSMUSG00000000037                  True                  True   \n",
      "\n",
      "                   gene_symbol_FINAL  \n",
      "EID                                   \n",
      "ENSMUSG00000000001             Gnai3  \n",
      "ENSMUSG00000000003              Pbsn  \n",
      "ENSMUSG00000000028             Cdc45  \n",
      "ENSMUSG00000000031               H19  \n",
      "ENSMUSG00000000037             Scml2  \n",
      "dict_keys(['BETA-DATA_hc_gene_programs_parsed_colors', 'BETA-DATA_hc_gene_programs_parsed_order', 'BETA-DATA_leiden_r1.5_parsed_colors', 'BETA-DATA_leiden_r1.5_parsed_order', 'cell_type_integrated_v2_colors', 'cell_type_integrated_v2_parsed_order', 'field_descriptions', 'study_colors', 'study_order', 'study_parsed_colors', 'study_parsed_order'])\n",
      "KeysView(AxisArrays with keys: BETA-DATA_X_umap, BETA-DATA_X_umap_opt, X_integrated, X_umap)\n",
      "<Compressed Sparse Row sparse matrix of dtype 'float64'\n",
      "\twith 1 stored elements and shape (5, 5)>\n",
      "  Coords\tValues\n",
      "  (0, 0)\t0.7450058814879538\n",
      "{'X': 'Expression normalized with integration-based size factors and transformed with log(x+1)', 'obs': {'*_high': 'Where * denotes hormone name. Do cells have high expression of the given hormone (ins, gsg, sst, ppy), determined per sample', '*_score': 'Where * denotes hormone name. Hormone scores computed per sample.', 'BETA-DATA_hc_gene_programs': 'Beta cell fine subtype annotation on integrated atlas. Fine annotation aimed at capturing all biollogically distinct beta cell subtypes (assesed based on gene program activity patterns).', 'BETA-DATA_hc_gene_programs_parsed': 'As BETA-DATA_hc_gene_programs, but with pretty names.Abbreviations: D-inter. - diabetic intermediate, NOD-D - NOD diabetic, M/F - male/female, chem - chem dataset, imm. - immature, lowQ - low quality.', 'BETA-DATA_leiden_r1.5': 'Clustering used to define beta cell coarse subtype annotation.', 'BETA-DATA_leiden_r1.5_parsed': 'Beta cell coarse subtype annotation on integrated atlas. Coarse annotation based on metadata information. Abbreviations: NOD-D - NOD diabetic, M/F - male/female, chem - chem dataset, imm. - immature, lowQ - low quality, hMT - high mitochondrial fraction', 'BETA-DATA_leiden_r1.5_parsed_const': 'As BETA-DATA_leiden_r1.5_parsed, but without pretty names. Used as an unchanging reference for coarse cell states.', 'BETA-DATA_leiden_r20': 'Cell clusters used to define beta-cell pseudobulk in some analyses.', 'CXG-DATA_GEO_accession': 'GEO accession of each dataset', 'CXG-DATA_age_approxDays': 'Approximate mapping of age column to days for the purpose of visualisation', 'CXG-DATA_cell_subtype_endothelial_reannotatedIntegrated': 'Endothelial cell subtype reannotation on integrated atlas based on known markers', 'CXG-DATA_cell_subtype_immune_reannotatedIntegrated': 'Immune cell subtype reannotation on integrated atlas based on known markers', 'CXG-DATA_chemical_stress': 'Application of chemicals to islets', 'CXG-DATA_diabetes_model': 'Diabetes model and any diabetes treatment', 'CXG-DATA_doublet_score': 'Scrublet doublet scores computed per sample; higher - more likely doublet', 'CXG-DATA_emptyDrops_LogProb_scaled': 'Log probability that droplet is empty computed per sample with emptyDrops and scaled to [0,1] per sample; higher - more likely empty droplet', 'CXG-DATA_log10_n_counts': 'log10(N counts)', 'CXG-DATA_mt_frac': 'Fraction of mitochondrial genes expression', 'CXG-DATA_n_genes': 'Number of expressed genes', 'CXG-DATA_sex_annotation': 'Was sex known from sample metadata (ground-truth) or was it determined bsed on Y-chromosomal gene expression (data-driven)', 'age': 'Age as defined in original publication, E - embryonic days, d - postnatal days, m - postnatal months, y - postnatal years', 'cell_filtering': 'FACS sorting', 'cell_subtype': 'Per-dataset annotation of cell types and for beta cells cell states that are annotated based on known markers. Used for integartion evaluation. Unnanotated cells start with NA.', 'cell_subtype_multiplet': 'As cell_subtype, but with potential multiplet cell clusters set to a single annotation that can be jointly filtered out.', 'cell_type': 'Per-dataset annotation of cell types used for integartion evaluation. Unnanotated cells start with NA.', 'cell_type_integrated_v2': 'Cell type reannotation on the integrated atlas.', 'cell_type_integrated_v2_parsed': 'As cell_type_integrated_v2, but with pretty names.Abbreviations: E - embryonic, endo. - endocrine, \"+\"\" symbol - likely doublet, prolif. - proliferative, lowQ - low quality, stellate a./q. - stellate activated/quiescent', 'cell_type_multiplet': 'As cell_type, but with potential multiplet cell clusters set to a single annotation that can be jointly filtered out.', 'cell_type_parsed': 'As cell_type, but with pretty names.', 'design': 'Brief sample description that gives information on differences between samples within dataset', 'file': 'Technical sample, in some cases equal to biological sample', 'g1_cyclone': 'G1 score of cyclone', 'g2m_cyclone': 'G2M score of cyclone', 'low_q': 'False for cells asigned to low quality clusters', 'phase_cyclone': 'Phase of cell cycle (cyclone)', 'pre_cell_type_original': 'Cell types as reported in the studies that generated the datasets', 'pre_cell_type_unified': 'Cell types as reported in the studies that generated the datasets; manually unified to a common naming scheme. Abberviations: E - embryonic, EP - endocrine progenitor/precursor, Fev+ - Fev positive, prolif. - proliferative, \"+\"\" symbol - likely doublet', 'reference': 'Was dataset used in initial integration hyperparameter evaluation, see reproducibility code for more information.', 's_cyclone': 'S score of cyclone', 'sex': 'Cell sex.', 'size_factors_integrated': 'Size factors computed on integrated embedding.', 'size_factors_sample': 'Size factors computed per sample.', 'strain': 'Mouse strain and genetic background', 'study': 'Dataset name.', 'study_parsed': 'As study, but with pretty name.', 'study_sample': 'Concatentation of columns with sample study and sample (file) information', 'study_sample_design': 'Concatentation of columns with sample study, sample (file), and design information', 'technique': 'Sequencing protocol', 'tissue': 'Tissue type'}, 'obsm': {'BETA-DATA_X_umap': 'Beta-cell specific UMAP embedding computed with default Scanpy parameters.', 'BETA-DATA_X_umap_opt': 'As BETA-DATA_X_umap, but with optimised parameters to enable better visualisation of cell clusters.', 'X_integrated': 'Integrated atlas embedding.', 'X_umap': 'Atlas-wide UMAP embedding.'}, 'raw.X': 'Raw expression counts', 'readme': 'The field description dictionary contains adata field explanations. Fields prefixed with \"BETA-DATA\" come from beta-cell specific adata object (in code saved in data_rawnorm_integrated_analysed_beta_v1s1_sfintegrated.h5ad) and fields prefixed with CXG-DATA come from adata prepared for cellxgene; other fields come from atlas-wide objects (in code saved in data_rawnorm_integrated_annotated.h5ad and data_integrated_analysed.h5ad). Expression contains all genes from count matrices of individual datasets and we specify in the below described var columns which genes were used for integration and atlas exploration. If a gene was missing from a dataset its expression was set to 0 for that dataset.', 'uns': {'*_colors': 'Colors for the column corresponding to *.', '*_order': 'Order of categories in the column corresponding to *.'}, 'var': {'CXG-DATA_feature_is_filtered': 'False for features not present in all datasets. This defines features used for atlas exploration.', 'CXG-DATA_present_*': '* Dataset name. True if feature was present in a dataset.', 'gene_symbol': 'Gene symbols from BioMart V103.', 'gene_symbol_FINAL': 'Gene symbol assigned for each gene first based on BioMart V103,  and if missing based on gene symbols from alignment information of datasets.', 'gene_symbol_original_matched': 'Gene symbols shared across annotations of individual datasets, some of which differ due to different genomic versions.If not shared set to np.nan.', 'used_integration': 'True if feature was used for integration.'}}\n"
     ]
    }
   ],
   "source": [
    "print(adata.shape)                      # (n_cells, n_genes) = (301796, 31706)\n",
    "print(adata)                            # adata 주요 구성 내용\n",
    "print(adata.obs.head())                 # 각 셀의 메타데이터: sample_id, condition, diabetes_status, etc.\n",
    "print(adata.var.head())                 # 유전자 정보\n",
    "print(adata.uns.keys())                 # 시각화용 설정값, 필드 설명 등\n",
    "print(adata.obsm.keys())                # PCA, UMAP, latent space 등 low-dimensional embedding\n",
    "print(adata.X[:5, :5])                  # 처음 5x5 행렬 출력\n",
    "print(adata.uns[\"field_descriptions\"])  # 메타데이터 항목 설명"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "791182b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cell_type_integrated_v2_parsed\n",
      "beta             102143\n",
      "alpha             40935\n",
      "immune            31703\n",
      "E non-endo.       29177\n",
      "delta             24775\n",
      "stellate a.       18332\n",
      "endothelial       13469\n",
      "ductal             8742\n",
      "E endo.            7748\n",
      "gamma              6999\n",
      "beta+delta         5104\n",
      "stellate q.        4970\n",
      "alpha+delta        1901\n",
      "beta+gamma         1209\n",
      "delta+gamma        1069\n",
      "endo. prolif.       887\n",
      "lowQ                853\n",
      "alpha+beta          683\n",
      "schwann             617\n",
      "acinar              480\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# beta-cell subset\n",
    "print(adata.obs['cell_type_integrated_v2_parsed'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "96721f37",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found 102143 beta cells\n"
     ]
    }
   ],
   "source": [
    "# beta cell 이름만 먼저 추출\n",
    "beta_cell_names = adata.obs_names[adata.obs['cell_type_integrated_v2_parsed'] == 'beta']\n",
    "print(f\"Found {len(beta_cell_names)} beta cells\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "044540af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sampling (RAM 폭발 방지)\n",
    "# 샘플링 크기 결정\n",
    "n_sample = 5000 if len(beta_cell_names) >= 5000 else len(beta_cell_names)\n",
    "# 무작위 샘플링\n",
    "np.random.seed(42)\n",
    "sampled_cells = np.random.choice(beta_cell_names, size=n_sample, replace=False)\n",
    "# sampling\n",
    "beta_cells_subset = adata[sampled_cells].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "371d2319",
   "metadata": {},
   "outputs": [],
   "source": [
    "# beta-cell subset + T1D/control 조건 라벨 정보\n",
    "beta_cells_subset.obs['condition'] = beta_cells_subset.obs['CXG-DATA_diabetes_model']\n",
    "beta_cells_subset.write(\"C:/Users/user/Desktop/T1D/beta_cells_condition_analysis.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "e82df237",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 5000 × 31706\n",
      "    obs: 'study_sample', 'study', 'file', 'reference', 'size_factors_sample', 'phase_cyclone', 's_cyclone', 'g2m_cyclone', 'g1_cyclone', 'sex', 'ins_score', 'ins_high', 'gcg_score', 'gcg_high', 'sst_score', 'sst_high', 'ppy_score', 'ppy_high', 'cell_filtering', 'age', 'strain', 'tissue', 'technique', 'study_sample_design', 'cell_type', 'cell_type_multiplet', 'cell_subtype', 'cell_subtype_multiplet', 'design', 'size_factors_integrated', 'pre_cell_type_unified', 'pre_cell_type_original', 'study_parsed', 'cell_type_parsed', 'low_q', 'BETA-DATA_leiden_r1.5', 'BETA-DATA_leiden_r20', 'BETA-DATA_hc_gene_programs', 'BETA-DATA_hc_gene_programs_parsed', 'BETA-DATA_leiden_r1.5_parsed', 'BETA-DATA_leiden_r1.5_parsed_const', 'CXG-DATA_n_genes', 'CXG-DATA_mt_frac', 'CXG-DATA_doublet_score', 'CXG-DATA_log10_n_counts', 'CXG-DATA_age_approxDays', 'CXG-DATA_cell_subtype_immune_reannotatedIntegrated', 'CXG-DATA_cell_subtype_endothelial_reannotatedIntegrated', 'CXG-DATA_emptyDrops_LogProb_scaled', 'CXG-DATA_diabetes_model', 'CXG-DATA_chemical_stress', 'CXG-DATA_GEO_accession', 'CXG-DATA_sex_annotation', 'cell_type_integrated_v2', 'cell_type_integrated_v2_parsed', 'condition'\n",
      "    var: 'gene_symbol', 'used_integration', 'gene_symbol_original_matched', 'CXG-DATA_feature_is_filtered', 'CXG-DATA_present_Fltp_2y', 'CXG-DATA_present_Fltp_adult', 'CXG-DATA_present_Fltp_P16', 'CXG-DATA_present_NOD', 'CXG-DATA_present_NOD_elimination', 'CXG-DATA_present_spikein_drug', 'CXG-DATA_present_embryo', 'CXG-DATA_present_VSG', 'CXG-DATA_present_STZ', 'gene_symbol_FINAL'\n",
      "    uns: 'BETA-DATA_hc_gene_programs_parsed_colors', 'BETA-DATA_hc_gene_programs_parsed_order', 'BETA-DATA_leiden_r1.5_parsed_colors', 'BETA-DATA_leiden_r1.5_parsed_order', 'cell_type_integrated_v2_colors', 'cell_type_integrated_v2_parsed_order', 'field_descriptions', 'study_colors', 'study_order', 'study_parsed_colors', 'study_parsed_order'\n",
      "    obsm: 'BETA-DATA_X_umap', 'BETA-DATA_X_umap_opt', 'X_integrated', 'X_umap'\n",
      "컬럼 목록: ['study_sample', 'study', 'file', 'reference', 'size_factors_sample', 'phase_cyclone', 's_cyclone', 'g2m_cyclone', 'g1_cyclone', 'sex', 'ins_score', 'ins_high', 'gcg_score', 'gcg_high', 'sst_score', 'sst_high', 'ppy_score', 'ppy_high', 'cell_filtering', 'age', 'strain', 'tissue', 'technique', 'study_sample_design', 'cell_type', 'cell_type_multiplet', 'cell_subtype', 'cell_subtype_multiplet', 'design', 'size_factors_integrated', 'pre_cell_type_unified', 'pre_cell_type_original', 'study_parsed', 'cell_type_parsed', 'low_q', 'BETA-DATA_leiden_r1.5', 'BETA-DATA_leiden_r20', 'BETA-DATA_hc_gene_programs', 'BETA-DATA_hc_gene_programs_parsed', 'BETA-DATA_leiden_r1.5_parsed', 'BETA-DATA_leiden_r1.5_parsed_const', 'CXG-DATA_n_genes', 'CXG-DATA_mt_frac', 'CXG-DATA_doublet_score', 'CXG-DATA_log10_n_counts', 'CXG-DATA_age_approxDays', 'CXG-DATA_cell_subtype_immune_reannotatedIntegrated', 'CXG-DATA_cell_subtype_endothelial_reannotatedIntegrated', 'CXG-DATA_emptyDrops_LogProb_scaled', 'CXG-DATA_diabetes_model', 'CXG-DATA_chemical_stress', 'CXG-DATA_GEO_accession', 'CXG-DATA_sex_annotation', 'cell_type_integrated_v2', 'cell_type_integrated_v2_parsed', 'condition']\n",
      "condition\n",
      "T2D_db/db-treated_VSG                      474\n",
      "T2D_db/db-treated_PairFeed                 462\n",
      "T2D_db/db                                  260\n",
      "T1D_NOD                                    162\n",
      "T1D_NOD_prediabetic                         98\n",
      "T2D_mSTZ-treated_estrogen                   84\n",
      "T2D_mSTZ-treated_GLP-1_estrogen             83\n",
      "T2D_mSTZ-treated_GLP-1_estrogen+insulin     68\n",
      "T2D_mSTZ                                    66\n",
      "T2D_mSTZ-treated_insulin                    56\n",
      "T2D_mSTZ-treated_GLP-1                      55\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# [확인] beta_cells_condition_analysis.h5ad 데이터 구조\n",
    "adata_beta_cond = sc.read(\"C:/Users/user/Desktop/T1D/beta_cells_condition_analysis.h5ad\")\n",
    "print(adata_beta_cond)\n",
    "print(\"컬럼 목록:\", adata_beta_cond.obs.columns.tolist())\n",
    "print(adata_beta_cond.obs[\"condition\"].value_counts())  # T1D vs control 분포"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a9de251c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "EID                                            ENSMUSG00000000001  \\\n",
      "index                                                               \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination            0.745006   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination            0.000000   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination            0.000000   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                          0.000000   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination            0.000000   \n",
      "\n",
      "EID                                            ENSMUSG00000000003  \\\n",
      "index                                                               \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                 0.0   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                 0.0   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                 0.0   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                               0.0   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                 0.0   \n",
      "\n",
      "EID                                            ENSMUSG00000000028  \\\n",
      "index                                                               \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                 0.0   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                 0.0   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                 0.0   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                               0.0   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                 0.0   \n",
      "\n",
      "EID                                            ENSMUSG00000000031  \\\n",
      "index                                                               \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                 0.0   \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                 0.0   \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                 0.0   \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                               0.0   \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                 0.0   \n",
      "\n",
      "EID                                            ENSMUSG00000000037  \n",
      "index                                                              \n",
      "CAAGATCGTCCAGTTA-1-SRR7610301-NOD_elimination                 0.0  \n",
      "GATGAAAGTTGTCGCG-1-SRR7610298-NOD_elimination                 0.0  \n",
      "AGTCTTTAGGAGCGTT-1-SRR7610301-NOD_elimination                 0.0  \n",
      "CTTCCTTGTACCCAGC-1-MUC13640-VSG                               0.0  \n",
      "CATCAAGAGATTACCC-1-SRR7610296-NOD_elimination                 0.0  \n"
     ]
    }
   ],
   "source": [
    "# expression 값을 DataFrame처럼 보기\n",
    "df = pd.DataFrame(adata.X[:5, :5].todense() if issparse(adata.X) else adata.X[:5, :5],\n",
    "                  index=adata.obs_names[:5], columns=adata.var_names[:5])\n",
    "print(df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
